File: project_euler/problem_531/solution.cpp
#include <bits/stdc++.h>
using namespace std;

typedef long long ll;
typedef __int128 lll;

/*
 * Problem 531: Chinese Leftovers
 *
 * Compute S = sum of g(phi(n), n, phi(m), m) over 1000000 <= n < m <= 1005000.
 * g(a,n,b,m) = smallest non-negative x with x=a(mod n), x=b(mod m), or 0.
 *
 * Method: Euler totient sieve + extended CRT via Bezout's identity.
 * Time: O(N log log N + K^2 log N).
 */

ll extended_gcd(ll a, ll b, ll &x, ll &y) {
    if (a == 0) { x = 0; y = 1; return b; }
    ll x1, y1;
    ll g = extended_gcd(b % a, a, x1, y1);
    x = y1 - (b / a) * x1;
    y = x1;
    return g;
}

ll crt(ll a, ll n, ll b, ll m) {
    ll u, v;
    ll g = extended_gcd(n, m, u, v);
    if ((a - b) % g != 0) return 0;
    ll mg = m / g;
    ll diff = ((lll)(b - a) / g % mg * u % mg + mg) % mg;
    ll lcm = n / g * m;
    ll x = (lll)a + (lll)n * diff;
    x = ((x % lcm) + lcm) % lcm;
    return x;
}

int main() {
    const int LO = 1000000, HI = 1005000;
    vector<ll> phi(HI + 1);
    iota(phi.begin(), phi.end(), 0);
    for (int i = 2; i <= HI; i++) {
        if (phi[i] == i) {
            for (int j = i; j <= HI; j += i)
                phi[j] -= phi[j] / i;
        }
    }

    ll S = 0;
    for (int n = LO; n <= HI; n++)
        for (int m = n + 1; m <= HI; m++)
            S += crt(phi[n], n, phi[m], m);

    cout << S << endl;
    return 0;
}

File: project_euler/problem_531/solution.py
"""
Problem 531: Chinese Leftovers

Compute S = sum of g(phi(n), n, phi(m), m) over 1000000 <= n < m <= 1005000,
where g(a,n,b,m) is the smallest non-negative solution to the system
x = a (mod n), x = b (mod m), or 0 if none exists.

Method: Extended CRT via Bezout's identity; totient sieve for phi values.
"""

from math import gcd

def extended_gcd(a, b):
    """Return (g, x, y) such that a*x + b*y = g = gcd(a, b)."""
    if a == 0:
        return b, 0, 1
    g, x, y = extended_gcd(b % a, a)
    return g, y - (b // a) * x, x

def crt(a, n, b, m):
    """Smallest non-negative x with x = a (mod n), x = b (mod m), or 0."""
    d = gcd(n, m)
    if (a - b) % d != 0:
        return 0
    lcm = n // d * m
    _, u, _ = extended_gcd(n, m)
    x = (a + n * ((b - a) // d) * u) % lcm
    return x

def euler_totient_sieve(limit):
    """Compute phi(k) for k = 0..limit via the Euler product sieve."""
    phi = list(range(limit + 1))
    for i in range(2, limit + 1):
        if phi[i] == i:  # i is prime
            for j in range(i, limit + 1, i):
                phi[j] -= phi[j] // i
    return phi

LO, HI = 1000000, 1005000
phi = euler_totient_sieve(HI)

S = 0
for n in range(LO, HI + 1):
    for m in range(n + 1, HI + 1):
        S += crt(phi[n], n, phi[m], m)

print(S)
